pkg/provider/azure_utils.go

```diff
@@ -447,36 +447,28 @@ func getResourceByIPFamily(resource string, isDualStack, isIPv6 bool) string {
 }
 
 // isFIPIPv6 checks if the frontend IP configuration is of IPv6.
-func (az *Cloud) isFIPIPv6(service *v1.Service, fip *network.FrontendIPConfiguration, isInternal bool) (bool, error) {
-	if isInternal {
-		if fip.FrontendIPConfigurationPropertiesFormat != nil {
-			if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion != "" {
-				return fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion == network.IPv6, nil
-			}
-			if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress != nil {
-				return net.ParseIP(*fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress).To4() == nil, nil
-			}
+func (az *Cloud) isFIPIPv6(rg string, fip *network.FrontendIPConfiguration) (bool, error) {
+	// First check private
+	// For an external LB, the FIP's private IP address is empty. The logic will
+	// go through this part and check public IP address then.
+	if fip.FrontendIPConfigurationPropertiesFormat != nil {
+		if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion != "" {
+			return fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddressVersion == network.IPv6, nil
+		}
+		if fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress != nil {
+			return net.ParseIP(*fip.FrontendIPConfigurationPropertiesFormat.PrivateIPAddress).To4() == nil, nil
 		}
-		klog.Errorf("Checking IP Family of frontend IP configuration %q of internal Service but "+
-			"it is not clear. It's considered to be IPv4",
-			pointer.StringDeref(fip.Name, ""))
-		return false, nil
 	}
-	var fipPIPID string
+
+	// Then Check public
 	if fip.FrontendIPConfigurationPropertiesFormat != nil && fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress != nil {
-		fipPIPID = pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
-	}
-	pipResourceGroup := az.getPublicIPAddressResourceGroup(service)
-	pips, err := az.listPIP(pipResourceGroup, azcache.CacheReadTypeDefault)
-	if err != nil {
-		return false, err
-	}
-	for _, pip := range pips {
-		id := pointer.StringDeref(pip.ID, "")
-		if !strings.EqualFold(fipPIPID, id) {
-			continue
+		fipPIPID := pointer.StringDeref(fip.FrontendIPConfigurationPropertiesFormat.PublicIPAddress.ID, "")
+		fipPIPName := fipPIPID[strings.LastIndex(fipPIPID, "/")+1:]
+		pip, exist, err := az.getPublicIPAddress(rg, fipPIPName, azcache.CacheReadTypeDefault)
+		if err != nil {
+			return false, err
 		}
-		if pip.PublicIPAddressPropertiesFormat != nil {
+		if exist && pip.PublicIPAddressPropertiesFormat != nil {
```

Comment:
Is it possible that the ipv6 public IP has not yet been created?

Reply:
It won't. One evidence is that the warning in `isFIPIPv6()` shouldn't appear in CCM log: "Checking IP Family of frontend IP configuration %q but it is not clear. It's considered to be IPv4".